structured/structured.proto

```diff
@@ -58,57 +54,39 @@ message ColumnType {
   repeated string vals = 4;
 }
 
-message Column {
-  optional string name = 1 [(gogoproto.nullable) = false];
-  optional ColumnType type = 2 [(gogoproto.nullable) = false];
-  optional bool nullable = 3 [(gogoproto.nullable) = false];
-}
-
-message Index {
-  optional string name = 1 [(gogoproto.nullable) = false];
-  optional bool unique = 2 [(gogoproto.nullable) = false];
-}
-
-message TableSchema {
-  message IndexByName {
-    optional Index index = 1 [(gogoproto.nullable) = false, (gogoproto.embed) = true];
-    // An ordered list of column names of which the index is comprised. Each
-    // column_name refers to a column in the TableSchema's columns.
-    repeated string column_names = 2;
-  }
-  optional Table table = 1 [(gogoproto.nullable) = false, (gogoproto.embed) = true];
-  repeated Column columns = 2 [(gogoproto.nullable) = false];
-  // An ordered list of indexes included in the table. The first index is the
-  // primary key; it is required.
-  repeated IndexByName indexes = 3 [(gogoproto.nullable) = false];
-}
-
 message ColumnDescriptor {
-  optional uint32 id = 1 [(gogoproto.nullable) = false, (gogoproto.customname) = "ID"];
-  optional Column column = 2 [(gogoproto.nullable) = false, (gogoproto.embed) = true];
+  optional string name = 1 [(gogoproto.nullable) = false];
+  optional uint32 id = 2 [(gogoproto.nullable) = false, (gogoproto.customname) = "ID"];
+  optional ColumnType type = 3 [(gogoproto.nullable) = false];
+  optional bool nullable = 4 [(gogoproto.nullable) = false];
 }
 
 message IndexDescriptor {
-  optional uint32 id = 1 [(gogoproto.nullable) = false, (gogoproto.customname) = "ID"];
-  optional Index index = 2 [(gogoproto.nullable) = false, (gogoproto.embed) = true];
-  // An ordered list of column ids of which the index is comprised. Each
-  // column_id refers to a column in the TableDescriptor's columns; special
-  // care is taken to update this when deleting columns.
-  repeated uint32 column_ids = 3 [(gogoproto.customname) = "ColumnIDs"];
+  optional string name = 1 [(gogoproto.nullable) = false];
+  optional uint32 id = 2 [(gogoproto.nullable) = false, (gogoproto.customname) = "ID"];
+  optional bool unique = 3 [(gogoproto.nullable) = false];
+  // An ordered list of column names of which the index is comprised. This list
+  // parallels the column_ids list. If duplicating the storage of the column
+  // names here proves to be prohibitive, we could clear this field before
+  // saving and reconstruct it after loading.
+  repeated string column_names = 4;
+  // An ordered list of column ids of which the index is comprised. This list
+  // parallels the column_names list.
+  repeated uint32 column_ids = 5 [(gogoproto.customname) = "ColumnIDs"];
```

Comment:
Just putting the name and id together doesn't insure we have both. We also have to check that the name and id are both present and that the name and id are consistent with an actual column. `TableDescriptor.Validate` does that checking. That said, It might be worth putting the two together for a minor cleanliness benefit. Let me see what that looks like.